### ISSUE TITLE
qx.core.Id API

### ISSUE THREAD

**Issue by cboulanger** [qooxdoo/qooxdoo#10033]
Picking up from https://github.com/qooxdoo/qooxdoo/pull/10030#pullrequestreview-477646177 , this issue can collect thoughts on how best to express the conceptual goals of `qx.core.Id` / `qx.core.MObjectId` via its API.

I am copying over the relevant parts from @johnspackman's and my thoughts on the issue:

@johnspackman :
In this review, I've mentioned a really useful "coding pattern" that object ID mechanism enables - here's some work that @cboulanger has done which illustrate the clarity

[Before, without QxObject ID](https://github.com/cboulanger/bibliograph/blob/91602806be0902fc6e4fad7eba894b8577544a41/src/client/bibliograph/source/class/bibliograph/ui/main/Toolbar.js) 

[After rewrite based on QxObjectID](https://github.com/cboulanger/bibliograph/blob/develop/src/client/bibliograph/source/class/bibliograph/ui/main/Toolbar.js)

Note how in the "before" the code has to hand craft the dependency order of objects and has to expose methods for each object, whereas the "after" focuses on declaring the widgets - and in this way the exposed widgets become part of an API.

Also, much like `@childcontrol` it's our intention to add JSDoc comment support for this, eg:

```
/**
  * @objectId(“toolbar”, qx.ui.toolbar.ToolBar) the main toolbar for the window
  */
_createObjectImpl(id) {
  switch(id) {
    /** @objectId(qx.ui.toolbar.Button) the OK button
    case “btnOk”:
```

[@johnspackman on renameing `_createObjectImpl` to `createObject`, which I had proposed since the name is so unsuited for general consumption]:

IMHO this is not a good way to change the API method name because it's non-obvious - the reader would normally expect the list of methods in `members` to define the API, and it is documented that the `_createQxObjectImpl` will be called, yet this change means that if the class *or any of it's descendant classes* implement `createQxObject` then the documented method is ignored.

I do think that in retrospect it would be good to change the naming here, because as @cboulanger says the object ID mechanism supports a coding pattern which is really useful in creating clear and concise widget UIs - more on that later, we've talked about a medium.com blog posting as well as docs to explain our thoughts.

IMHO the "create" method should stay protected though, because part of the API specification is that "create" will only be called once for each ID (and it could easily cause issues if that was not followed).  By making it protected and using the public `getQxObject` to control access, the `getQxObject` is a gatekeeper which reliably enforces those API rules.

I think I would prefer that `_createQxObjectImpl` and `_createQxObject` are simply swapped around, i.e. that the method which people override is called `_createQxObject` and the internal method which `getQxObject` defers to is `_createQxObjectImpl`.

This would be a breaking API change, but as we're talking about pre-release there is room to decide to do this; IMHO the benefits are better clarity because the design pattern that it promotes (or: that we intend to promote) is really very effective.

[@johnspackman on attaching "ownees" to other "owners" than the current object:]

[A]lthough we have built it so that it is possible to add objects by ID, that's really intended for situations where the object does not implement `_createQxObjectImpl`, ie it's a work around for existing code.  The workaround means that objects can be located via a global ID string (eg "dlgLoginForm/edtUsername"), the initial use case was for automated testing.

The coding pattern which is so useful is centered around an object creating it's own objects, and they are exposed kind of like an API.  For example, a login form dialog box might expose objects with the IDs "edtUsername", "edtPassword", and "btnOk" and document them as two text fields and a button; there will be a lot more widgets used to make up that form but these are not necessarily exposed (or possibly the IDs could be prefixed with "_" or "__").  

Preserving the Owner/Ownee relationship is critical because it allows name conflicts to be avoided - but when we promote adding objects to another objects, we're breaking that protection because we are adding the new object directly into the namespace of the `owner`.  

This could have some use, in that you would be able to override the creation of a widget (eg provide the `edtUsername` textfield of a login dialog) **but** this will only work if the object has not yet been created by the dialog.  If the dialog is later modified in such a way that the constructor calls `getQxObject("edtUsername")`, then the calling code will immediately break.  The correct, and the only reliable, way to change the implementation of a field would be to derive a class from the login dialog and provide your own implementation of `_createQxObjectImpl`.

Also, if we consider the IDs an API, then the object has the right to define new objects and change or remove protected/private objects as needed; but this solution will add additional, owned objects into the namespace of an object on the fly.  Even putting name conflicts to one side for a moment, I'm not convinced that this is the right way either - it feels quite unstructured, and the equivalent of adding methods to an object on the fly.  I feel that the expected way to add functionality would be by deriving a class, and that would mean that the additions form part of the code structure (as opposed to runtime) and are much better understood and documented - you could for example easily detect all classes which derive and then check their `_createQxObjectImpl` and then know *for a fact* what API object IDs are valid in that object.

AIUI the intention of this change is to support compositing of objects, ie assist building the object graph - but I think that the correct way to do that is to either use the owner/ownee ID mechanism, where you can have "grand-owner/owner/ownee/etc" or to have a class which is specifically intended for compositing in that way.  That special "compositing" class would be the place to support a lot of other features which are very relevant to compositing (such as recording modified state, persistence, etc).

@cboulanger :

As you say, the method names of this API are not ideal, in particular not very good in expressing the philosophy behind the system. If the priority is, as you say, that the hierarchy of owner/ownee is best not to be fiddled with (as in my attempt to attach an "ownee" to a different owner than the one producing it), `getQxObject` or `_createQxObject` doesn't really convey that idea. (`addOwnedQxObject` is better, but from the perspective of the design, a workaround, not a central part of the API). No immediate alternatives come to my mind, but maybe others have an idea. I wouldn't switch `_createQxObject` and `_createQxObjectImpl`, but rather find better semantics to replace them. 

I wonder if the idea behind `qx.core.Id` could be better expressed with "idChild"/"idParent" or "logicalChild"/"logicalParent" rather than with the owner/ownee or "qxObject" vocabulary. "parent" and "child" are well understood concepts in tree structures and the specific nature of the qx.core.Id API is that it is not about a layout hierarchy but a logical hierarchy.

**Comment by cboulanger** [comment]
One could also image this API to be completely different: if the methods were not part of a mixin, but a mixin gave acesss to a separate object which then provided the API, this would allow a terse style and avoid API clutter. Something like 

```javascript
obj.qxId().getId();
obj.qxId().setId("foo");
obj.qxId().getChild("bar"); // or obj.qxId().getOwnee("bar"); 
obj.qxId().getParent(); // or obj.qxId().getOwner();
```

This could still rely on an implementation method on the main object,

**Comment by johnspackman** [comment]
Originally we used `getObjectId()` and this broke existing code because it conflicted with existing developer code, so the "qx" prefix was a quick fix to work around that.  But the reality is that adding any features into a class like `qx.core.Object` has a significant risk of naming conflict, which is constraining for us as core developers, so rather than add `qxId()` would it be better to implement `qx()`, so that any Qooxdoo-specific functionality can be attached, rather than "just" ID specific methods?

**Comment by cboulanger** [comment]
I like that idea very much. It would be good if `qx.core.Object` had as little extra methods as possible. Since we can use environment variables to control what is available in the API, we can make this fully BC.

**Comment by cboulanger** [comment]
And in one's own code, a few search/replace operation would be enough to migrate to the new API

**Comment by cboulanger** [comment]
If one looks at https://qooxdoo.org/qxl.apiviewer/#qx.core.Object , the ammount of `assertX` methods is really off-puting. They are available only in `qx.debug`  mode, aren't they? Apart from that, there would be some candidates for cleanup.

**Comment by johnspackman** [comment]
yes, I think that was a mistake putting them in there TBH, but nothing we can do about it now.  However, I do think that it would help a lot if API viewer did not show mixin methods in the class, and also did not show automatically get/set/reset methods for each property.

**Comment by cboulanger** [comment]
Yes, but some mixins provide essential functionality, whereas others really aren't interesting. I guess this would have to be determined manually - a list of mixins which should only be shown on demand when checking a menu checkbox. The "show/hide mixins / inherited mixins" options don't really help because they don't make the distinction between essential and non-essential stuff. 

Getters and setters should be hidden by default.
